### PR TITLE
add support for `*.tar` and `*.tar.gz` archives

### PR DESCRIPTION
**Pull request by bennettp123**
Hi there! Thanks for providing this GitHub action—I really like the colorized output. ✨ 

This PR adds support for archives in `.tar` or `.tar.gz` format.

* usage: `archive: path/to/archive.tar.gz`
* file type is recognised using file extension: `.tar` or `.tar.gz`
* no change to behaviour of `directory`: if specified, it still creates an archive using `zip`

### CODE REVIEW

**Review by iBotPeaches — COMMENTED** [review]
Thanks for the PR. Left some comments and working to fix CI pipeline for pull requests.

File: deploy.sh [hunk 1]
```diff
@@ -82,10 +82,31 @@ else
     ZIP_FILENAME="$INPUT_ARCHIVE"
 fi
 
+if [ -z "$BUNDLE_TYPE" ]; then
+    # permitted values for BUNDLE_TYPE can be found here: https://docs.aws.amazon.com/codedeploy/latest/userguide/application-revisions-push.html#push-with-cli
+    case "$ZIP_FILENAME" in
+        *.tar)
+            BUNDLE_TYPE=tar
```
[the remark this answers is not in the document]
Comment (bennettp123) [reply deploy.sh:89]: Thanks @iBotPeaches, that should be fixed now.

File: deploy.sh [hunk 2]
```diff
@@ -82,10 +82,31 @@ else
     ZIP_FILENAME="$INPUT_ARCHIVE"
 fi
 
+if [ -z "$BUNDLE_TYPE" ]; then
```
[the remark this answers is not in the document]
Comment (bennettp123) [reply deploy.sh:85]: You're right, thanks. I was considering allowing it to be overridden, and then forgot to add it to action.yaml. What do you think?

File: deploy.sh [hunk 3]
```diff
@@ -82,10 +82,31 @@ else
...
+            # assume it's a zipfile
+            BUNDLE_TYPE=zip
+            ;;
+    ecas
+fi
 
-if [ "$(unzip -l "$ZIP_FILENAME" | grep -q appspec.yml)" = "0" ]; then
-    echo "::error::$ZIP_FILENAME was not generated properly (missing appspec.yml)."
-    exit 1;
+if [ $BUNDLE_TYPE == 'zip' ]; then
```
[the remark this answers is not in the document]
Comment (bennettp123) [reply deploy.sh:101]: Yeah interesting.

I should point out that I ran the listing manually, and it also didn't warn about the missing quotes—either because it's not set from untrusted input, or because the default settings of the linter don't warn on this kind of issue.

I agree that it's probably best to quote it—especially if it's sourced from action.yaml.

File: deploy.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (iBotPeaches) [reply deploy.sh:85]: If you are reading the suffix, not sure you need it. Otherwise enters strange situation where you give a zip, but specify a tar.

File: deploy.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (bennettp123) [reply deploy.sh]: Done

File: deploy.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (bennettp123) [reply deploy.sh:99]: I've added the input parameter to action.yaml, and also added some input validation

File: deploy.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (bennettp123) [reply deploy.sh]: (weirdly, the linter also didn't complain about the misspelled `ecas` statement!)

File: action.yaml [hunk 4]
```diff
@@ -28,6 +28,10 @@ inputs:
   archive:
     description: 'Zip to be used for deployment, instead of archiving a directory.'
     required: false
+  bundle_type:
+    description: >
+      The type of archive: tar, tgz or zip. By default, the extension will
+      be used to auto-detect this value.
```
Comment (iBotPeaches) [inline action.yaml:34]: I don't think we should add this parameter. It clouds the list and if we auto-detect it anyway - then doesn't seem needed. Especially since adding this can produce a situation where the archive is a zip, but was told it was a tar.

Less is more for this action so its easily auditable.

File: action.yaml [hunk 5]
```diff
@@ -40,6 +44,9 @@ inputs:
   codedeploy_group:
     description: 'AWS CodeDeploy Application Group'
     required: true
+  codedeploy_config_name:
+    description: 'Override the AWS CodeDeploy configuration name'
+    required: false
```
Comment (iBotPeaches) [inline action.yaml:49]: Not following this change - you should make this a separate pull request so we don't mix the bundle type issue with this.

File: deploy.sh [hunk 6]
```diff
@@ -188,18 +228,24 @@ pollForActiveDeployments
 
 # 5) Poll / Complete
 function deployRevision() {
+    EXTRA_ARGS=''
+    if [ -n "$INPUT_CODEDEPLOY_CONFIG_NAME" ]; then
+        EXTRA_ARGS="--deployment-config-name $INPUT_CODEDEPLOY_CONFIG_NAME"
+    fi
+
```
Comment (iBotPeaches) [inline deploy.sh:235]: Lets split this out to a different pr.

**Review by iBotPeaches — CHANGES_REQUESTED** [review]

File: deploy.sh (hunk 6 above)
Comment (bennettp123) [reply deploy.sh:217]: Yep, my bad, this should definitely be a separate PR.

Comment (bennettp123) [reply deploy.sh]: Done 😁

**Review by iBotPeaches — APPROVED** [review]

**Review by Vacoris — APPROVED** [review]

**Comment by iBotPeaches** [comment]
thanks for prs!
